File: examples/Fortran/openmp_converted/GameOfLife_solution_omp.f90
program gol
    implicit none
    integer :: rows, cols, generations
    integer, allocatable :: world(:,:), old_world(:,:)
    integer :: g
    integer :: long
    character(len=:), allocatable :: arg
    integer :: i
    integer, dimension(3) :: n
    ! Read cmd line args
    DO i=1, COMMAND_ARGUMENT_COUNT()
       CALL GET_COMMAND_ARGUMENT(NUMBER=i, LENGTH=long)
       ALLOCATE(CHARACTER(len=long) :: arg)
       CALL GET_COMMAND_ARGUMENT(NUMBER=i, VALUE=arg)
       READ(arg,'(i10)') n(i)
       DEALLOCATE(arg)
    END DO
    rows=n(1)
    cols=n(2)
    generations=n(3)
    allocate(world(0:rows+1,0:cols+1), old_world(0:rows+1,0:cols+1))
    old_world(:,:) = 0
    call fill_world
    !$ACC enter data copyin(world, old_world)
!$omp target enter data map(to:world,old_world)
    do g=1,generations
        call save_world(rows,cols, world(:,:), old_world(:,:))
        call next(rows,cols, world(:,:), old_world(:,:))
        print *, "Cells alive at generation ", g, ": ", alive(rows, cols, world)
        call output_world("generation", g, world(:,:), rows, cols)
    enddo
    !$ACC exit data delete(world, old_world)
!$omp target exit data map(delete:world,old_world)
    deallocate(world, old_world)
    contains
    integer function alive(rows, cols, world) result(cells)
        implicit none
        integer, intent(in) :: rows, cols
        integer, intent(in) :: world(0:rows+1,0:cols+1)
        integer :: r,c
        cells = 0
        !$ACC parallel loop reduction(+:cells)
!$omp target teams loop reduction(+:cells)
        do r=1, rows
            do c=1, cols
               cells = cells + world(r,c) 
            enddo
        enddo
        end function alive
    subroutine save_world(rows,cols,world, old_world)
        implicit none
        integer, intent(in) :: rows,cols
        integer, intent(in) :: world(0:rows+1,0:cols+1)
        integer, intent(out) :: old_world(0:rows+1,0:cols+1)
        integer :: r,c
        !$ACC parallel loop
!$omp target teams loop
        do r=1, rows
            do c=1, cols
                old_world(r,c) = world(r,c)
            enddo
        enddo
    end subroutine save_world

    subroutine next(rows,cols,world, old_world)
        implicit none
        integer, intent(in) :: rows,cols
        integer, intent(in) :: old_world(0:rows+1,0:cols+1)
        integer, intent(out) :: world(0:rows+1,0:cols+1)
        integer :: r,c
        integer :: neigh
        !$ACC parallel loop
!$omp target teams loop
        do r=1, rows
            do c=1, cols
        neigh = old_world(r-1,c-1)+old_world(r,c-1)+old_world(r+1,c-1)+&
                old_world(r-1,c)+old_world(r+1,c)+&
                old_world(r-1,c+1)+old_world(r,c+1)+old_world(r+1,c+1)
                if (old_world(r,c) == 1 .and. (neigh<2.or.neigh>3) ) then
                    world(r,c) = 0
                else if (neigh == 3) then
                    world(r,c) = 1
                endif
            enddo
        enddo
    end subroutine next

    subroutine fill_world
        implicit none
        integer :: r,c, temp
        real*8 :: test
        do r=1,rows
            do c=1,cols
                call random_number(test)
                temp = mod(floor(test*100),4)
                if (temp.eq.0) then 
                    temp = 1
                else
                    temp = 0
                endif
                world(r,c) = temp
            enddo
        enddo
    end subroutine fill_world
    
    subroutine output_world(name, g, world, rows, cols)
        implicit none
        character(len=*), intent(in) :: name
        character(len=1024) :: filename
        integer, intent(in) :: g
        integer, intent(in) :: rows,cols
        integer, intent(in) :: world(0:rows+1,0:cols+1)
        integer :: r,c
        integer :: my_unit
        integer(kind=1) :: output(0:rows+1,0:cols+1)
        write(filename,"(A,I5.5,A)") trim(name), g,".gray"
        print *, trim(filename)
        !$ACC update self(world)
!$omp target update from(world)
        do r=0,rows+1
            do c=0,cols+1
                output(r,c) = mod(world(r,c),2)*127
            enddo
        enddo
        open(newunit=my_unit, file=filename, access="stream", form="unformatted")
        write(my_unit, pos=1) output
        close(my_unit)
    end subroutine output_world
end program gol

! Code was translated using: intel-application-migration-tool-for-openacc-to-openmp GameOfLife_solution.f90

File: examples/C/openmp_converted/GameOfLife_solution_omp.c
#include <stdio.h>
#include <stdlib.h>

void output_world(int* restrict world, int rows, int cols, int generation)
{
    /**
     * Write a file with the world inside
     * @param world: a pointer to the storage for the current step
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border
     */
    char* path = (char*) malloc(sizeof(char)*80);
    sprintf(path, "generation%05d.gray", generation);
    FILE* f = fopen(path, "wb");
    unsigned char* mat = (unsigned char*) malloc(sizeof(unsigned char)*(rows+2)*(cols+2));
    #pragma acc parallel loop copyout(mat[(rows+2)*(cols+2)]) present(world[(rows+2)*(cols+2)])
#pragma omp target teams loop map(from:mat[(rows+2)*(cols+2)]) map(present,\
            alloc:world[(rows+2)*(cols+2)])
    for (int i=0; i<rows+2; ++i)
        for (int j=0; j<cols+2; ++j)
            mat[i*cols+j] = (unsigned char) world[i*cols+j] * 255;
    fwrite(world, sizeof(unsigned char), (rows+2)*(cols+2), f);
    fclose(f);
}
void next(int* restrict world, int* restrict  oworld, int rows, int cols)
{
    /**
     * Apply the rules and compute the next generation
     * @param world: a pointer to the storage for the current step
     * @param oworld: a pointer to the storage for the previous step
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border
     */
    int neigh = 0;
    int row_current = 0;
    int row_above = 0;
    int row_below = 0;
#pragma acc parallel loop present(world[:(rows+2)*(cols+2)], oworld[:(rows+2)*(cols+2)])
#pragma omp target teams loop map(present,alloc:world[:(rows+2)*(cols+2)],\
            oworld[:(rows+2)*(cols+2)])
    for (int r=1; r<=rows; ++r)
        for (int c=1; c<=cols; ++c)
        {
            row_current = r*(cols+2);
            row_above = (r-1)*(cols+2);
            row_below = (r+1)*(cols+2);
            neigh = oworld[row_above + c-1] + oworld[row_above + c]   + oworld[row_above + c+1] +
                    oworld[row_current + c+1]+                        oworld[row_current + c-1] + 
                    oworld[row_below + c-1] + oworld[row_below + c] + oworld[row_below + c+1];
            if (oworld[r*(cols+2)+c] == 1 && (neigh<2||neigh>3))
                world[r*(cols+2)+c] = 0;
            else if (neigh==3)
                world[r*(cols+2)+c] = 1;
        }
} 

void save(int* restrict world, int* restrict oworld, int rows, int cols)
{
    /**
     * Save the current world to oworld
     * @param world: a pointer to the storage for the current step
     * @param oworld: a pointer to the storage for the previous step
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border
     */
#pragma acc parallel loop collapse(2) present(world[:(rows+2)*(cols+2)], oworld[:(rows+2)*(cols+2)])
#pragma omp target teams loop collapse(2) map(present,\
            alloc:world[:(rows+2)*(cols+2)],oworld[:(rows+2)*(cols+2)])
    for (int r=1; r<=rows; ++r)
        for (int c=1; c <= cols; ++c)
            oworld[r*(cols+2) + c] = world[r*(cols+2) + c];
}

int alive(int* restrict world, int rows, int cols)
{
    /**
     * Compute the number of cells alive at the current generation
     * @param world: a pointer to the storage for the current step
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border
     */
    int cells = 0;
#pragma acc parallel loop collapse(2) reduction(+:cells) present(world[:(rows+2)*(cols+2)])
#pragma omp target teams loop reduction(+:cells) collapse(2) map(present,\
            alloc:world[:(rows+2)*(cols+2)])
    for (int r=1; r <= rows; ++r)
        for (int c=1; c <= cols; ++c)
            cells += world[r*(cols+2) + c];
    return cells;
}

void fill_world(int* restrict world, int rows, int cols)
{
    /**
     *  Set the initial state of the world
     * @param world: a pointer to the storage for the current step
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border
     */
    for (int r=1; r <= rows; ++r)
        for (int c=1; c <= cols; ++c)
            world[r*(cols+2) + c] = rand()%4==0 ?1 : 0;
    // The border of the world is a dead zone
    for (int i=0;i<=rows;++i)
    {
        world[i*(cols+2)] = 0;
        world[i*(cols+2)+cols+1] = 0;
    }
    for (int j=0; j<cols; ++j)
    {
        world[j] = 0;
        world[(rows+1)*(cols+2)+j] = 0;
    }
}

int* allocate(int rows, int cols)
{
    /**
     * Allocate memory for a 2D array
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border
     * @return a pointer to the matrix
     */
    
    int* mat = (int*) malloc((rows+2)*(cols+2)*sizeof(int));
#pragma acc enter data create(mat[0:(rows+2)*(cols+2)])
#pragma omp target enter data map(alloc:mat[0:(rows+2)*(cols+2)])
    
    return mat;
}

void destroy(int* mat, int rows, int cols)
{
    /**
     * Free memory for a 2D array
     * @param mat: a pointer to the matrix to free
     * @param rows: the number of rows without the border
     * @param cols: the number of columns without the border 
     */
#pragma acc exit data delete(mat[0:(rows+2)*(cols+2)])
#pragma omp target exit data map(delete:mat[0:(rows+2)*(cols+2)])
    free(mat);
}

int main(int argc, char** argv)
{
    int rows, cols, generations;
    int* world;
    int* oworld;

    if (argc < 4)
    {
        printf("Wrong number of arguments: Please give rows cols and generations\n");
        return 1;
    }
    rows = strtol(argv[1], NULL, 10);
    cols = strtol(argv[2], NULL, 10);
    generations = strtol(argv[3], NULL, 10);
    
    world = allocate(rows, cols);
    oworld = allocate(rows, cols);
    fill_world(world, rows, cols);
    printf("Initial state set\n");
#pragma acc update device(world[0:(rows+2)*(cols+2)])
#pragma omp target update to(world[0:(rows+2)*(cols+2)])
    printf("Cells alive at generation %d: %d\n", 0, alive(world, rows, cols));
    for (int g=1; g <= generations; ++g)
    {
        save(world, oworld, rows, cols);
        next(world, oworld, rows, cols);
        output_world(world, rows, cols, g);
        printf("Cells alive at generation %4d: %d\n", g, alive(world, rows, cols));
    }

    destroy(world, rows, cols);
    destroy(oworld, rows, cols);

    return 0;
}

// Code was translated using: /home/very/bin/acc2mp -async=nowait -specify-language=C -generate-report GameOfLife_solution.c
